### PR TITLE
HCAP-1415 - Bulk Graduate

### PR DESCRIPTION
**Pull request by mmumby**
## Work completed
- Updated existing API - `/post-hire-status` to accept an array of `participantIds` instead of a single ID. 
- Updated the FE  cohort-details page to include 'Bullk Graduate' button, multi select on the cohort page, participant table - visible for HA
- If the form is being rendered by "bulk graduate' disable the 'unsuccessful flow' and only allow graduation to be set.
- If a participant already has a graduation status, disallow the user from opening the form, show Alert above the table (See screenshot)

UI: 
Screen Shot 2023-03-17 at 2 30 20 PM
Screen Shot 2023-03-20 at 2 21 35 PM
Screen Shot 2023-03-20 at 4 53 47 PM

modal/form if setting the status on an individual
Screen Shot 2023-03-17 at 2 31 40 PM

### CODE REVIEW

File: server/services/cohorts.ts
Comment (mmumby) [inline server/services/cohorts.ts:98]: This was causing a 500 error locally if logged in as HA visiting `cohorts/:id`

File: client/src/components/fields/RenderRadioGroup.js
Comment (rhattie-ey) [inline client/src/components/fields/RenderRadioGroup.js:48]: What is the purpose of this check? Why might `options` contain a falsy value?

File: client/src/components/modal-forms/ManageGraduationForm.js
Comment (rhattie-ey) [inline client/src/components/modal-forms/ManageGraduationForm.js:14]: This is a container with text inside, correct? Generally I'd prefer to see em or rem used here if that's the case. It's a nitpick though.

File: server/routes/post-hire-status.ts
Comment (rhattie-ey) [inline server/routes/post-hire-status.ts:31]: Both of these arrays should ideally have typing, but especially notParticipants.

File: server/routes/post-hire-status.ts
Comment (rhattie-ey) [inline server/routes/post-hire-status.ts:45]: Could potentially use a template here like in the logger error, to make it clearer what went wrong from the client end (may help in Cypress debugging, for example).

File: server/routes/post-hire-status.ts
Comment (rhattie-ey) [inline server/routes/post-hire-status.ts:37]: This can be made cleaner and more purely functional with a filter instead of `push`.

```suggestion
    const notParticipants = (
      await Promise.all(participantIds.map(async (id) => ({id, valid: await getParticipantByID(id)).length > 0)})
    ).filter((participant) => participant.valid);
```

File: server/routes/post-hire-status.ts
Comment (rhattie-ey) [inline server/routes/post-hire-status.ts:55]: As with participants, this can be simplified with a filter. See my other suggestion for an implementation example.

File: server/services/cohorts.ts
Comment (rhattie-ey) [inline server/services/cohorts.ts:97]: Some of this optional chaining could be avoided by adding typing to `cohortParticipants`, but this also works.

File: client/src/components/fields/RenderRadioGroup.js
Comment (mmumby) [reply client/src/components/fields/RenderRadioGroup.js:48]: In ManageGraduationForm.js I conditionally rendered the options depending on if bulk graduation was occurring (We only want to allow the successful flow) 
```
 options={[
                    {
                      value: postHireStatuses.postSecondaryEducationCompleted,
                      label: 'Graduated',
                    },
                    !isBulkGraduate && {
                      value: postHireStatuses.cohortUnsuccessful,
                      label: 'Unsuccessful cohort',
                    },
                  ]}
```
This returned options as [object, false], which was displaying the radio button without anything else.

File: server/routes/post-hire-status.ts
Comment (mmumby) [reply server/routes/post-hire-status.ts:37]: The push was partly due to the error message getting logged, I wanted to inform which id's didn't have participants
`Participant(s) do not exist with id ${JSON.stringify(notParticipants)}`

**Review by vanics — DISMISSED** [review]
It looks good!

@rhattie-ey provided a few nice suggestions that you might want to consider, but overall is a sound implementation.

File: client/src/components/modal-forms/ManageGraduationForm.js
Comment (vanics) [reply client/src/components/modal-forms/ManageGraduationForm.js:14]: That's an interesting suggestion. I personally tend not to use `em` and `rem` as much but when talking about accessibility they make way more sense. 

As a side note for reference, @mmumby if you consider to follow up on this nitpick consider that:
EM is relative to the parent element’s font size, so if you wish to scale the element’s size based on its parent’s size, use EM.

REM is relative to the root (HTML) font size, so if you wish to scale the element’s size based on the root size, no matter what the parent size is, use REM. If you’ve used EM and are finding sizing issues due to lots of nested elements, REM will probably be the better choice.

My recommendation would be to be consistent with other parts of the application, so whatever we use the most in a specific context ideally would be the way to go.

Comment (mmumby) [reply client/src/components/modal-forms/ManageGraduationForm.js]: Great discussion - I removed the Alert from this form, and no longer needed this class! I will keep this in mind for sizing in the future though.

File: server/routes/post-hire-status.ts
Comment (mmumby) [reply server/routes/post-hire-status.ts]: This example doesn't work as anticipated - We cannot chain a filter onto a Promise.all(), we would require a second variable to filter/map to get the correct format for the alert. 

```
const notParticipants: {
      id: string;
      valid: boolean;
    }[] = await Promise.all(
      participantIds.map(async (id) => ({ id, valid: (await getParticipantByID(id)).length > 0 }))
    );

    const filteredParticipants: {
      id: string;
      valid: boolean;
    }[] = notParticipants.filter((obj) => !obj.valid);

    if (filteredParticipants.length) {
      logger.error({
        action: 'post-hire-status_post',
        message: `Participant(s) do not exist with id **${JSON.stringify(
          filteredParticipants.map(({ id }) => id)
        )}**`,
      });

      return res
        .status(422)
        .send(
          `Participant(s) do not exist with id **${JSON.stringify(
            filteredParticipants.map(({ id }) => id)
          )}.** Please check participant ID`
        );
    }
    ```
    
 Additionally, I can handle the .map in the `filteredParticipants` const vs in JSON.stringify - Although I think this refactor as a whole does not simplify the previous implementation.

File: client/src/components/fields/RenderRadioGroup.js
Comment (rhattie-ey) [reply client/src/components/fields/RenderRadioGroup.js:48]: In that case, I'd say you're better off filtering out the potentially false element there (`.filter(Boolean)` on the array as you define it) instead of making this logic expect arbitrary falsy values mixed in with actual options.

File: server/routes/post-hire-status.ts
Comment (rhattie-ey) [reply server/routes/post-hire-status.ts]: Ideally `participantIds` should also have typing, but at least giving `notParticipants` a type is a big improvement.

File: client/src/components/fields/RenderRadioGroup.js
Comment (mmumby) [reply client/src/components/fields/RenderRadioGroup.js]: Updated!

File: server/routes/post-hire-status.ts
Comment (mmumby) [reply server/routes/post-hire-status.ts]: Updated!

File: server/routes/post-hire-status.ts
Comment (mmumby) [reply server/routes/post-hire-status.ts]: Discussed Via teams chat. Updated.

File: server/routes/post-hire-status.ts
Comment (rhattie-ey) [inline server/routes/post-hire-status.ts:27]: Add an ESLint directive here to prevent errors, potentially alongside a JSDoc comment at the top of the type to explain what this is used for (and by extension why it would need a snake case property name).

File: server/routes/post-hire-status.ts
Comment (rhattie-ey) [reply server/routes/post-hire-status.ts]: Looks good!

**Review by rhattie-ey — COMMENTED** [review]
One minor change to prevent ESLint errors, and this is otherwise approved by me. Great work!

File: server/routes/post-hire-status.ts
Comment (mmumby) [reply server/routes/post-hire-status.ts]: Updated!

**Review by rhattie-ey — APPROVED** [review]
Looks good, although I'd suggest addressing the comments before merging. I'll re-approve if needed.

File: server/routes/post-hire-status.ts
Comment (fw-noel) [inline server/routes/post-hire-status.ts:53]: Bit interesting here. I think it works fine, though would've been nice to have a service to just get multiple participants by ids.

**Review by fw-noel — APPROVED** [review]

File: server/routes/post-hire-status.ts
Comment (mmumby) [reply server/routes/post-hire-status.ts:30]: Updated in the follow up testing PR!

**Pull request merged**
